bitpacker/src/blocked_bitpacker.rs

```diff
@@ -0,0 +1,146 @@
+        let encoded = offset | (num_bits as u64) << 56;
+        Self { encoded }
+    }
+    fn offset(&self) -> u64 {
+        (self.encoded << 8) >> 8
+    }
+    fn num_bits(&self) -> u8 {
+        (self.encoded >> 56) as u8
+    }
+}
+#[test]
+fn metadata_test() {
+    let meta = BlockedBitpackerEntryMetaData::new(50000, 6);
+    assert_eq!(meta.offset(), 50000);
+    assert_eq!(meta.num_bits(), 6);
+}
+
+impl BlockedBitpacker {
+    pub fn new() -> Self {
+        let mut compressed_blocks = vec![];
+        compressed_blocks.resize(8, 0);
+        Self {
+            compressed_blocks,
+            cache: vec![],
+            offset_and_bits: vec![],
+        }
+    }
+
+    pub fn get_memory_usage(&self) -> usize {
+        self.compressed_blocks.capacity()
+            + self.offset_and_bits.capacity()
+                * std::mem::size_of_val(&self.offset_and_bits.get(0).cloned().unwrap_or_default())
+            + self.cache.capacity()
+                * std::mem::size_of_val(&self.cache.get(0).cloned().unwrap_or_default())
+    }
+
+    pub fn add(&mut self, val: u64) {
+        self.cache.push(val);
+        if self.cache.len() == BLOCK_SIZE as usize {
+            self.flush();
+        }
+    }
+
+    pub fn flush(&mut self) {
+        if self.cache.is_empty() {
+            return;
+        }
+        let mut bit_packer = BitPacker::new();
+        let num_bits_block = self
+            .cache
+            .iter()
+            .map(|val| compute_num_bits(*val))
+            .max()
+            .unwrap();
+        self.compressed_blocks
+            .resize(self.compressed_blocks.len() - 8, 0); // remove padding for bitpacker
+        let offset = self.compressed_blocks.len() as u64;
+        for val in self.cache.iter() {
+            bit_packer
+                .write(*val, num_bits_block, &mut self.compressed_blocks)
+                .expect("cannot write bitpacking to output"); // write to im can't fail
+        }
+        bit_packer.flush(&mut self.compressed_blocks).unwrap();
+        self.offset_and_bits
+            .push(BlockedBitpackerEntryMetaData::new(offset, num_bits_block));
+
+        self.cache.clear();
+        self.compressed_blocks
+            .resize(self.compressed_blocks.len() + 8, 0); // add padding for bitpacker
+    }
+    pub fn get(&self, idx: usize) -> u64 {
+        let metadata_pos = idx / BLOCK_SIZE as usize;
+        let pos_in_block = idx % BLOCK_SIZE as usize;
+        if let Some(metadata) = self.offset_and_bits.get(metadata_pos) {
+            let unpacked = BitUnpacker::new(metadata.num_bits()).get(
+                pos_in_block as u64,
+                &self.compressed_blocks[metadata.offset() as usize..],
+            );
+            unpacked
+        } else {
+            self.cache[pos_in_block]
+        }
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = u64> + '_ {
+        let bitpacked_elems = self.offset_and_bits.len() * BLOCK_SIZE;
+        let iter = (0..bitpacked_elems)
```

Comment:
maybe add a todo to optimize this?

Reply:
yes, compression can also be further optimized